File: src/main/java/com/example/finsentinel/service/trading/uta/CcxtBroker.java
package com.example.finsentinel.service.trading.uta;

import com.example.finsentinel.service.trading.engine.AccountInfo;
import com.example.finsentinel.service.trading.engine.CcxtTradingEngine;
import com.example.finsentinel.service.trading.engine.OrderRequest;
import com.example.finsentinel.service.trading.engine.OrderResult;
import com.example.finsentinel.service.trading.engine.PositionInfo;

import java.util.EnumSet;
import java.util.List;
import java.util.Objects;
import java.util.Set;

/**
 * IBroker adapter for CCXT (XChange) crypto exchanges.
 * Supports CRYPTO security type only.
 * Not a Spring bean — created by BrokerRegistry.
 */
public class CcxtBroker implements IBroker {

    private final CcxtTradingEngine engine;

    public CcxtBroker(CcxtTradingEngine engine) {
        this.engine = Objects.requireNonNull(engine, "CcxtTradingEngine must not be null");
    }

    @Override
    public String brokerId() {
        return "ccxt-" + engine.engineName();
    }

    @Override
    public String displayName() {
        return "CCXT (" + engine.engineName() + ")";
    }

    @Override
    public Set<SecurityType> supportedSecurityTypes() {
        return EnumSet.of(SecurityType.CRYPTO);
    }

    @Override
    public Set<BrokerCapability> capabilities() {
        return EnumSet.of(BrokerCapability.SPOT_TRADING, BrokerCapability.MARKET_DATA);
    }

    @Override
    public OrderResult placeOrder(Contract contract, OrderRequest request) {
        var nativeRequest = new OrderRequest(
                contract.toEngineSymbol(), request.side(), request.type(),
                request.qty(), request.notional(), request.price(),
                request.stopPrice(), request.timeInForce(), request.reduceOnly());
        return engine.placeOrder(nativeRequest);
    }

    @Override
    public List<PositionInfo> getPositions() {
        return engine.getPositions();
    }

    @Override
    public List<OrderResult> getOrders() {
        return engine.getOrders();
    }

    @Override
    public AccountInfo getAccount() {
        return engine.getAccount();
    }

    @Override
    public boolean cancelOrder(String orderId) {
        return engine.cancelOrder(orderId);
    }
}

File: apps/api/src/trading/brokers/ccxt.broker.ts
import { SecurityType, BrokerCapability, Contract } from '@finsentinel/shared';
import type { IBroker } from '../interfaces/broker';
import type { CcxtTradingEngine } from '../engines/ccxt-trading.engine';
import type {
  OrderRequest,
  OrderResult,
  PositionInfo,
  AccountInfo,
  MarketClock,
} from '../interfaces/types';

/**
 * CCXT broker capabilities — spot trading only (crypto exchanges).
 */
const CCXT_CAPABILITIES: Set<BrokerCapability> = new Set([BrokerCapability.SPOT_TRADING]);

/**
 * CcxtBroker — wraps CcxtTradingEngine with Contract-aware symbol conversion.
 *
 * NOT an Injectable. Created by BrokerRegistry at runtime.
 *
 * CCXT broker adapter:
 * - Supports CRYPTO security type only
 * - Converts contract to CCXT-native symbol via Contract.toEngineSymbol()
 *   (e.g., BTC → BTC/USDT using contract.currency)
 * - Delegates all operations to the underlying CcxtTradingEngine
 */
export class CcxtBroker implements IBroker {
  constructor(private readonly _engine: CcxtTradingEngine) {}

  // -- Identity ---------------------------------------------------------------

  brokerId(): string {
    return 'ccxt';
  }

  displayName(): string {
    return 'CCXT (Crypto)';
  }

  // -- Capability checks ------------------------------------------------------

  supportedSecurityTypes(): Set<SecurityType> {
    return new Set([SecurityType.CRYPTO]);
  }

  capabilities(): Set<BrokerCapability> {
    return CCXT_CAPABILITIES;
  }

  canHandle(contract: Contract): boolean {
    return contract.secType === SecurityType.CRYPTO;
  }

  // -- Order execution --------------------------------------------------------

  async placeOrder(contract: Contract, request: OrderRequest): Promise<OrderResult> {
    const adapted: OrderRequest = {
      ...request,
      symbol: contract.toEngineSymbol(),
    };

    return this._engine.placeOrder(adapted);
  }

  // -- Delegated operations ---------------------------------------------------

  async getPositions(): Promise<PositionInfo[]> {
    return this._engine.getPositions();
  }

  async getOrders(): Promise<OrderResult[]> {
    return this._engine.getOrders();
  }

  async getAccount(): Promise<AccountInfo> {
    return this._engine.getAccount();
  }

  async cancelOrder(orderId: string): Promise<boolean> {
    return this._engine.cancelOrder(orderId);
  }

  async syncOrders(): Promise<OrderResult[]> {
    return this._engine.syncOrders!();
  }

  async getMarketClock(): Promise<MarketClock> {
    return this._engine.getMarketClock!();
  }

  async searchContracts(_query: string): Promise<Contract[]> {
    return [];
  }

  // -- Engine access ----------------------------------------------------------

  /** Expose the underlying engine for direct access when needed. */
  engine(): CcxtTradingEngine {
    return this._engine;
  }
}
